experimental/examples/token/Token.idr

```diff
@@ -36,30 +43,91 @@ getAccountBalance address accounts = case lookup address accounts of
+||| @storage the current storage
+total updateAllowance : (allower : Address) -> (allowee : Address) -> (tokens : Nat) -> (storage : Storage) -> Storage
+updateAllowance allower allowee tokens storage =
+  case tokens of
+    Z => record
+           {accounts =
+              insert -- update allower's allowance map
+              allower -- k
+              (MkAccount -- v
+                (getAccountBalance allower (accounts storage)) -- balance of allower is unchanged
+                (delete allowee (getAccountAllowance allower (accounts storage))) -- delete allowee from allower's allowance map
+              )
+              (accounts storage) -- update the accounts map
+           } storage
+    n => record
+           {accounts =
+              insert
+              allower -- k
+              (MkAccount -- v
+                (getAccountBalance allower (accounts storage)) -- balance
+                (insert allowee n (getAccountAllowance allower (accounts storage))) -- allowance map
+              )
+              (accounts storage) -- accounts map
+           } storage
+
+||| approve update the allowance map of the caller of the contract
+||| @spender the address the caller approve the tokens to transfer to
+||| @tokens the amount of tokens approved to be transferred
+||| @storage the current storage
+total approve : (spender : Address) -> (tokens : Nat) -> (storage : Storage) -> Storage
+approve spender tokens storage =
+  let caller = owner storage in -- caller is the caller of the contract, it is set to be the owner for now.
+    updateAllowance caller spender tokens storage
+
+||| transferFrom can be called by anyone, transferring amount no larger than the approved amount
+||| @from the address the tokens to be transferred from
+||| @dest the address the tokens to transfer to
+||| @tokens the amount to be transferred
+||| @storage the current storage
+total transferFrom : (from : Address) -> (dest : Address) -> (tokens : Nat) -> (storage : Storage) -> Either Error Storage
+transferFrom from dest tokens storage =
+  case lookup dest (getAccountAllowance from (accounts storage)) of
+    Nothing => Left NotAllowedToSpendFrom
+    (Just allowed) =>
+      case lte tokens allowed of
+        False => Left NotEnoughAllowance
+        True => let updatedStorage = updateAllowance from dest (minus allowed tokens) storage in
+                    performTransfer from dest tokens updatedStorage
+
 ||| createAccount transfers tokens from the owner to an address
 ||| @dest the address of the account to be created
 ||| @tokens the amount of tokens in the new created account
+||| @storage the current storage
 total createAccount : (dest : Address) -> (tokens : Nat) -> (storage : Storage) -> Either Error Storage
 createAccount dest tokens storage =
   let owner = owner storage in
-      case owner == owner of --when sender can be detected, check sender == owner.
+      case owner == owner of --when caller can be detected, check caller == owner.
```

Comment:
In order to have a (fake) caller field, we can do something like this - https://github.com/cryptiumlabs/juvix/blob/develop/experimental/examples/token/Token.fst#L85 - where we have a transaction type including a data field with the call information and a calling address.

Reply:
> In order to have a (fake) caller field, we can do something like this - https://github.com/cryptiumlabs/juvix/blob/develop/experimental/examples/token/Token.fst#L85 - where we have a transaction type including a data field with the call information and a calling address.

Yes, I could do that. Although I think it may be better to have a `fakeLib` that I import instead of having that in the code? What do you think? Lmk I can do either.

Reply:
Sure, that seems like a good idea.